File: Week_09/680/valid_palindrome_ii.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1

        def helper(s, i, j):
            while i < j:
                if s[i] != s[j]:
                    return False
                i += 1
                j -= 1
            return True

        while left < right:
            if s[left] == s[right]:
                left += 1
                right -= 1
            else:
                return helper(s, left + 1, right) or helper(s, left, right - 1)
        return True

File: Week_09/680/valid_palindrome_ii.go
package main

func validPalindrome(s string) bool {
	left, right := 0, len(s)-1
	for left < right {
		if s[left] == s[right] {
			left++
			right--
		} else {
			leftFlag, rightFlag := true, true
			for i, j := left+1, right; i < j; i, j= i+1, j-1 {
				if s[i] != s[j] {
					leftFlag = false
					break
				}
			}
			for i, j := left, right-1; i < j; i, j = i+1, j-1 {
				if s[i] != s[j] {
					rightFlag = false
					break
				}
			}
			return leftFlag || rightFlag
		}
	}
	return true
}
